### PR TITLE
AF-1084: Perform simple indexing of filenames separate from full indexing that requires a build

### PR DESCRIPTION
**Pull request by mbarkley**

### CODE REVIEW

**Comment by mbarkley** [comment]
Related PRs:
https://github.com/kiegroup/appformer/pull/255
https://github.com/kiegroup/kie-wb-common/pull/1529
https://github.com/kiegroup/drools-wb/pull/833
https://github.com/kiegroup/optaplanner-wb/pull/265

**Comment by mbarkley** [comment]
Jenkins, execute full downstream build.

**Comment by mbarkley** [comment]
Jenkins, execute full downstream build.

**Comment by mbarkley** [comment]
@ederign @porcelli Here are the PRs for the indexing improvements. I'm sorry I wasn't able to send them earlier.

Ideally this PR would include tests for the [MultiIndexerLock](https://github.com/kiegroup/appformer/pull/255/files#diff-25861c77fb7d3a968e83ae201f5afd71) and the [ConstrainedIndexerScheduler](https://github.com/kiegroup/appformer/pull/255/files#diff-38efd51a8400a3b48a457714e5b5b3e3), but I am hoping that I can add tests for these in a PR next week after the feature freeze. As it stands these are indirectly tested through the multiple indexing tests in the platform.

**Comment by mbarkley** [comment]
Note that the ConstrainedIndexerScheduler supports dependencies between indexers, but it's not yet being used in the workbench. For now it only uses the priorities to schedule the LibraryIndexer before the rest.

File: uberfire-extensions/uberfire-metadata/uberfire-metadata-commons-io/src/main/java/org/uberfire/ext/metadata/io/BatchIndex.java [hunk 1]
```diff
@@ -91,151 +92,115 @@ public void run() {
...
+                    logInformation("Completed indexing of " + fs.getName());
+                    if (callback != null) {
+                        callback.run();
+                    }
+                })
+                .exceptionally(ex -> {
+                    try {
+                        throw ex;
+                    } catch (DisposedException de) {
+                        logWarning("Batch index couldn't finish. [@" + fs.getName() + "]");
```
Comment (jhrcek) [inline uberfire-extensions/uberfire-metadata/uberfire-metadata-commons-io/src/main/java/org/uberfire/ext/metadata/io/BatchIndex.java]: Does it make sense to add exception to log message here as well as it's done in following catch branches? Similar applies to other `logWarning` calls here..

Comment (mbarkley) [reply uberfire-extensions/uberfire-metadata/uberfire-metadata-commons-io/src/main/java/org/uberfire/ext/metadata/io/BatchIndex.java]: @jhrcek All of the warning messages are for situations where indexing is terminated before completion. This was the same behaviour that existed prior to the change-set.

Maybe these should be errors, but I don't think they have anything to do with the subject matter of this PR so I'd prefer to open a separate issue for this.

**Comment by mbarkley** [comment]
@jhrcek Here's the high level description:

Previously, when indexing was triggered (by either the `BatchIndex` or the watch service), we would run all the indexers on the entire batch of files in a single thread, and indexing would not be completed until this entire thing was finished. This had the negative effect of preventing the LibraryIndexer (used for showing the asset list) from completing until some of the slower indexers finished.

With this PR, when indexing is triggered, each indexer processes relevant files in its own thread (scheduled via the `ExecutorService`). Each indexer is able to commit its changes independently, and external services can query whether the index is ready for a particular indexer (see the `LibraryServiceImpl` in the related kie-wb-common PR). With these changes, the asset list is now available much faster for newly imported projects.

The scheduler that decides the order in which indexers run support priorities and dependencies. Indexers with a lower priority number are run first (currently just the LibraryIndexer -- see `DefaultApplicationScopedProducer` in the kie-wb-common PR). Eventually we will also use the dependencies so that some indexers do not start until another indexer has already completed.

Please let me know if you have any follow up questions.

**Comment by jhrcek** [comment]
@mbarkley thank you for nice explanation. I'll leave it up to @tomasdavidorg to review and test this (already working on it).

**Comment by mbarkley** [comment]
Jenkins, execute full downstream build.

**Comment by tomasdavidorg** [comment]
Hi @mbarkley, I tested war from this PR and I found some minor issues:

1. When you import example projects, the loading pop-ups ("Please wait") blink several times 
loading_import

2. After project(s) is imported, the window with project(s) to import remains open. It is not redirected to the space where project(s) should be imported.

3. If you clone a bigger repository and you open the project during indexing, error modal dialog is displayed. 
screenshot from 2018-03-23 14-04-27

Otherwise works good. Thanks :-).

**Comment by mbarkley** [comment]
@tomasdavidorg I tried reproducing these issues, but no luck. Note that I rebuilt my own war file for this freshly rebased off of master (of kie-wb-webapp in kie-wb-distributions).

1. When I import all samples, I see an "Importing..." popup, several seconds of waiting, and then the "Evaluation Process" project is loaded. I was actually able to observe the multiple "Please wait" messages on war file built from master a few days ago, so I think this was unrelated and possibly fixed by another commit.

2. As described above, importing all the samples redirects me to the "Evaluation Process" project page. I observe this both in the war file built from this PR and a war file built from master yesterday.

3. I did not see this error when importing the big project with my war file or the one from master (thank you for sharing that, by the way). That error means that the ProjectScreen encountered an error while being initialized, which suggests to me that possibly you didn't build all repositories and maybe some sources were not on the same change-set?

I can upload my war file for you if you'd like to confirm my results (since the downstream build will take some time to complete).

**Comment by tomasdavidorg** [comment]
That would be great if you could share the war with me. Thanks.

**Comment by mbarkley** [comment]
@tomasdavidorg @ederign Since this didn't make the Friday cut-off, I'll be taking today to rebase the PRs and add tests for the ConstrainedScheduler and the MultiIndexerLock. I will build a war for you @tomasdavidorg when these changes are ready.

**Comment by mbarkley** [comment]
Jenkins, execute full downstream build.

**Comment by mbarkley** [comment]
@ederign PR is ready for review. I've reabsed off master, added tests for the lock and scheduler, and fixed/simplified the scheduler logic by delegating more to CompletableFutures.

**Comment by mbarkley** [comment]
Jenkins, execute full downstream build.

**Comment by mbarkley** [comment]
Jenkins, execute full downstream build.

**Comment by mbarkley** [comment]
@tomasdavidorg I've restarted the downstream build because of some failures I think are unrelated, but [here](https://kie-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/appformer-downstream-pullrequests/67/artifact/downstream-repos/kie-wb-distributions/kie-wb-parent/kie-wb-distribution-wars/kie-wb/target/kie-wb-7.8.0-SNAPSHOT-wildfly11.war) is the war file from the last build (requires VPN). I tried it myself and I do not see the issues you previously described.

One thing I do see that I don't think happened before is that you now see `pom.xml`, `kmodule.xml`, and `project.repositories` files in the asset list. I'm not sure if that was the case before, but I think this will be fixed in a PR that @adrielparedes is working on to hide files from the asset list that aren't used by other indexers.

**Comment by adrielparedes** [comment]
@mbarkley yes, I'm about to create that PR to hide those files

**Comment by tomasdavidorg** [comment]
@mbarkley Thank you. The issues are gone. I didn't see any other issues except of the visible` pom.xml`, ... files.

**Comment by mbarkley** [comment]
Jenkins, execute full downstream build.

File: uberfire-extensions/uberfire-metadata/uberfire-metadata-api/src/main/java/org/uberfire/ext/metadata/engine/IndexerScheduler.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        /**
+         * @param jobsByIndexerId A map of jobs to run by their {@link Indexer} id. Must not be null.
+         * @return An {@link IndexerScheduler} for scheduling execution of the given jobs. Never null.
+         */
+        IndexerScheduler create(Map<String, ? extends Supplier<List<IndexEvent>>> jobsByIndexerId);
+    }
+
+    /**
+     * @param executor An {@link ExecutorService} used for scheduling any asynchronous jobs. Must not be null.
+     * @return A stream of {@link CompletableFuture CompleteableFutures} for all jobs scheduled. Never null.
```
Comment (ederign) [inline uberfire-extensions/uberfire-metadata/uberfire-metadata-api/src/main/java/org/uberfire/ext/metadata/engine/IndexerScheduler.java]: small typo

File: uberfire-extensions/uberfire-metadata/uberfire-metadata-commons-io/src/main/java/org/uberfire/ext/metadata/io/ConstrainedIndexerScheduler.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+                              .stream()
+                              .sorted(comparingInt(node -> node.priority))
+                              .map(node -> node.id)
+                              .map(id -> schedule(executor, memoizedJobs, id));
+    }
+
+    private CompletableFuture<Pair<String, List<IndexEvent>>> schedule(ExecutorService executor,
+                                                                       Map<String, CompletableFuture<Pair<String, List<IndexEvent>>>> memoizedJobs,
+                                                                       String id) {
+        if (memoizedJobs.containsKey(id)) {
```
[the remark this answers is not in the document]
Comment (mbarkley) [reply uberfire-extensions/uberfire-metadata/uberfire-metadata-commons-io/src/main/java/org/uberfire/ext/metadata/io/ConstrainedIndexerScheduler.java]: @ederign Correct. Would `createdJobs` be better?

[the remark this answers is not in the document]
Comment (ederign) [reply uberfire-extensions/uberfire-metadata/uberfire-metadata-commons-io/src/main/java/org/uberfire/ext/metadata/io/ConstrainedIndexerScheduler.java]: yep, but feel free to just ignore it.

**Pull request merged**